### ISSUE TITLE
Segfault in new_scope

### ISSUE THREAD

**Issue by tongson** [mruby/mruby#2769]
Found through afl-fuzz. Thousands of crashes, going to report obvious syntax errors.

VERSION AFFECTED: mruby 1.1.0

TESTCASE:
```
Proc.new do
  do
end
def x==x
end
```

BACKTRACE:
```
#0  0x000000000043c245 in new_scope (p=0x6fbbe0, body=0x6fc468) at src/parse.y:277
#1  0x000000000043e83f in yyparse (p=0x6fbbe0) at src/parse.y:1172
#2  0x000000000044ba11 in mrb_parser_parse (p=0x6fbbe0, c=0x6be9a0) at src/parse.y:5377
#3  0x000000000044bfbf in mrb_parse_file (mrb=0x6a2010, f=0x6dd7f0, c=0x6be9a0) at src/parse.y:5523
#4  0x000000000044c45d in mrb_load_file_cxt (mrb=0x6a2010, f=0x6dd7f0, c=0x6be9a0) at src/parse.y:5608
#5  0x00000000004024b9 in main (argc=2, argv=0x7fffffffe468) at /root/mruby-1.1.0/mrbgems/mruby-bin-mruby/tools/mruby/mruby.c:225
```

**Comment by cremno** [comment]
>Thousands of crashes

I guess they aren't all fixed by now, are they? Do you mind uploading the findings folder of afl to somewhere, so someone can investigate them? Plain Ruby code shouldn't cause any segmentation faults.

**Comment by cremno** [comment]
@matz: What about similar code (`new_module()`, `new_def()`)?

```ruby
module A
  Proc.new do
    def x==x end
  end
end
```
```ruby
def a
  Proc.new do
    def x==x end
  end
end
```
```diff
diff --git a/src/parse.y b/src/parse.y
index 336b20a..0571fe0 100644
--- a/src/parse.y
+++ b/src/parse.y
@@ -270,12 +270,17 @@ local_add(parser_state *p, mrb_sym sym)
   }
 }
 
+static node*
+locals_node(parser_state *p)
+{
+  return p->locals ? p->locals->car : NULL;
+}
+
 /* (:scope (vars..) (prog...)) */
 static node*
 new_scope(parser_state *p, node *body)
 {
-  node *n = p->locals ? p->locals->car : NULL;
-  return cons((node*)NODE_SCOPE, cons(n, body));
+  return cons((node*)NODE_SCOPE, cons(locals_node(p), body));
 }
 
 /* (:begin prog...) */
@@ -602,35 +607,35 @@ new_undef(parser_state *p, mrb_sym sym)
 static node*
 new_class(parser_state *p, node *c, node *s, node *b)
 {
-  return list4((node*)NODE_CLASS, c, s, cons(p->locals->car, b));
+  return list4((node*)NODE_CLASS, c, s, cons(locals_node(p), b));
 }
 
 /* (:sclass obj body) */
 static node*
 new_sclass(parser_state *p, node *o, node *b)
 {
-  return list3((node*)NODE_SCLASS, o, cons(p->locals->car, b));
+  return list3((node*)NODE_SCLASS, o, cons(locals_node(p), b));
 }
 
 /* (:module module body) */
 static node*
 new_module(parser_state *p, node *m, node *b)
 {
-  return list3((node*)NODE_MODULE, m, cons(p->locals->car, b));
+  return list3((node*)NODE_MODULE, m, cons(locals_node(p), b));
 }
 
 /* (:def m lv (arg . body)) */
 static node*
 new_def(parser_state *p, mrb_sym m, node *a, node *b)
 {
-  return list5((node*)NODE_DEF, nsym(m), p->locals->car, a, b);
+  return list5((node*)NODE_DEF, nsym(m), locals_node(p), a, b);
 }
 
 /* (:sdef obj m lv (arg . body)) */
 static node*
 new_sdef(parser_state *p, node *o, mrb_sym m, node *a, node *b)
 {
-  return list6((node*)NODE_SDEF, o, nsym(m), p->locals->car, a, b);
+  return list6((node*)NODE_SDEF, o, nsym(m), locals_node(p), a, b);
 }
 
 /* (:arg . sym) */
@@ -668,14 +673,14 @@ new_block_arg(parser_state *p, node *a)
 static node*
 new_block(parser_state *p, node *a, node *b)
 {
-  return list4((node*)NODE_BLOCK, p->locals->car, a, b);
+  return list4((node*)NODE_BLOCK, locals_node(p), a, b);
 }
 
 /* (:lambda arg body) */
 static node*
 new_lambda(parser_state *p, node *a, node *b)
 {
-  return list4((node*)NODE_LAMBDA, p->locals->car, a, b);
+  return list4((node*)NODE_LAMBDA, locals_node(p), a, b);
 }
 
 /* (:asgn lhs rhs) */
```

**Comment by matz** [comment]
@cremno the diff looks good to me. Could you submit a pull request? Or should I add by myself?

**Comment by cremno** [comment]
@matz: I'll create a PR very soon. I thought there might be a preferable solution.

**Comment by tongson** [comment]
The first two crashes I reported was found a few seconds after running afl-fuzz. After 3 cycles and almost a month of running it I now have 780+ unique crashes.

Who should I send these crashing test cases? It's only 44Kb compressed.

**Comment by carsonmcdonald** [comment]
@tongson Can you put them all in a repo and then provide a reference to them here? I was actually looking at the same thing a few days ago but only let afl-fuzz run for a couple hours. It came up with a couple of issues that I believe are the same as what you have found along with 200 "unique" crashes but they all are resolved by fixes for the initial couple it found. I would be curious if the fixes I have for those two will fix almost all or all of the ones you have from the longer run you did. I can run my changes on your results if you put them in a repo.

**Comment by cremno** [comment]
Yeah I ran it too for some hours but #2777 fixed most of the crashes I found.

Can you just post them here? Or is there a reason you rather not? Anyway I'll check them out. You could create a git repo or simply upload them to some file hosting site like https://pomf.se/. Thanks!

**Comment by tongson** [comment]
Uploaded to:
https://github.com/tongson/afl-fuzz-findings/tree/master/mruby-1.1.0

Only against 1.1.0. Most of the testcases are probably invalid now because of the recent fixes.

**Comment by cremno** [comment]
After merging #2814 we're down to 20 crashes. 

Oh, and [id:000737,sig:11,src:028869,op:havoc,rep:8
](https://github.com/tongson/afl-fuzz-findings/blob/master/mruby-1.1.0/crashes/id:000737%2Csig:11%2Csrc:028869%2Cop:havoc%2Crep:8) is a simple stack overflow (because of repetitive unary plus calls). I don't think there's much that can be done.

**Comment by cremno** [comment]
3! Only the stack overflow one and these two remain:
https://github.com/tongson/afl-fuzz-findings/blob/master/mruby-1.1.0/crashes/id:000519%2Csig:11%2Csrc:018483%2Cop:havoc%2Crep:1
https://github.com/tongson/afl-fuzz-findings/blob/master/mruby-1.1.0/crashes/id:000520%2Csig:11%2Csrc:018483%2Cop:havoc%2Crep:1

Might be a codegen issue.

**Comment by cremno** [comment]
```ruby
a = *[1], if false
    Class
    2
  end
```
```
    4 000 OP_ARRAY      R2      R2      0
    1 001 OP_LOADI      R3      1
    1 002 OP_ARRAY      R3      R3      1
    1 003 OP_ARYCAT     R2      R3
    1 004 OP_LOADF      R3
    1 005 OP_JMPNOT     R3      009
    # ...
    3 010 OP_ARYPUSH    R3      R4              ; R3 should be R2
    3 011 OP_MOVE       R1      R3              ; R1:a
    3 012 OP_STOP
```

**Comment by cremno** [comment]
Oh, the following two also fail (so now we're really down to 3):

https://github.com/tongson/afl-fuzz-findings/blob/master/mruby-1.1.0/crashes/id:000499%2Csig:06%2Csrc:017562%2Cop:havoc%2Crep:2
https://github.com/tongson/afl-fuzz-findings/blob/master/mruby-1.1.0/crashes/id:000544%2Csig:06%2Csrc:019390%2Cop:havoc%2Crep:2

The assertion in `src/gc.c:1100` fails. But it might be also a codegen issue.

**Comment by matz** [comment]
If fixed #2825.  Is the fix improve this too?

**Comment by cremno** [comment]
@matz: I think this can be closed. I couldn't reproduce any crashes found by @tongson with afl-fuzz as of 727f2485c9028cbd5eb4085e965bcaada6fb53f6.
